### ISSUE TITLE
YOLO V3 INT 8 quantization weights

### ISSUE THREAD

**Issue by YL-123** [AlexeyAB/yolo2_light#15]
Hi, thanks for the great work! I run the INT-8 inference using YoloV3 weights(248M) with the flag -quantized successfully, just wondering is it possible to save the INT 8 quantization weights first (size would be ~60M), and then run a float 32 inference, is anyone compare the performance difference here?

**Comment by AlexeyAB** [comment]
@Yinling-123 Hi,

If you will use INT8-weights and FLOAT32-inference:
* then **speed will be the same** as for FLOAT32-weights and FLOAT32-inference
* **accuracy will be slightly lower** than for FLOAT32-weights and FLOAT32-inference

So it doesn't make sense and isn't implemented, because FLOAT32-weights and FLOAT32-inference better.

In this repository is implemented INT8-weights and INT8-inference, +30% speedup and -2% accuracy.

**Comment by Yinling-123** [comment]
@AlexeyAB Thanks for your quick reply, I will be greatly appreciated if you can guide me how to converter the FLOAT32-weights to INT8-weights use this repo?

**Comment by Yinling-123** [comment]
Hi, is it possible to add the converter feature (which save the INT8 weights) in this repo, I found gplhegde version darknet has the converter but not support YOLO V3 weights,

**Comment by AlexeyAB** [comment]
@Yinling-123 Hi,
No, there is no code for saving INT8-weights currently.

**Comment by ambr89** [comment]
Yes, now you can!

**Comment by isgursoy** [comment]
> In this repository is implemented INT8-weights and INT8-inference, +30% speedup and -2% accuracy.
You mean fp32 weights and int8 inference right?

**Comment by AlexeyAB** [comment]
@isgursoy 

> You mean fp32 storage and int8 inference right?

I mean fp32 storage, fp32->int8 weights conversion, and int8 inference.

**Comment by isgursoy** [comment]
Why one would need to dump int8 like https://github.com/AlexeyAB/darknet/issues/1405 ? What are pros and cons doing that?
Thanks for your time.

**Comment by AlexeyAB** [comment]
@isgursoy 

> Why one would need to dump int8 (on Pascal or Tesla) like AlexeyAB/darknet#1405?

Can you clarify? I don't understand.

**Comment by isgursoy** [comment]
Hi,
Let's say I exported fp32 weights as you described in https://github.com/AlexeyAB/darknet/issues/1405.
I wonder the pros and cons of using int8 storage on Pascal or Tesla GPUs.

**Comment by AlexeyAB** [comment]
@isgursoy 
Pros - small weights-file size
Cons - you can't use this weights-file as pre-trained weights-file

**Comment by isgursoy** [comment]
Can you roughly estimate gpu mem gain and the loss of accuracy if there will be?
Motivation for us is accuracy-mem tradeoff here.

Regarding pretrained weights, it wont be problem if we get export after training right?

**Comment by AlexeyAB** [comment]
May be it will use ~3x less GPU_RAM

**Comment by isgursoy** [comment]
Without loss of accuracy?

**Comment by AlexeyAB** [comment]
If you just will store weights-int8 instead of weights-fp32 then you willn't get any gain in GPU-RAM / CPU-RAM and will not get any loss of accuracy.

To get less GPU-RAM usage you must implement all layers **int8** in yolov3: upsample, route, shortcut. And should use only `l.biases_gpu`, `l.weights_int8_int8x4_gpu` and `l.output_gpu_int8` instead of: https://github.com/AlexeyAB/yolo2_light/blob/master/src/additionally.c#L2820-L2827
You will get low drop in accuracy.

**Comment by w840401** [comment]
> 是的，現在可以了！

How can i use ?!
